plutus-core/plutus-ir/src/PlutusIR/Transform/LetFloatIn.hs

```diff
@@ -0,0 +1,406 @@
+  where
+    floatTermInner ::
+        Usages.Usages ->
+        Term tyname name uni fun a ->
+        Term tyname name uni fun (a, Uniques)
+    floatTermInner usgs = go
+      where
+        -- \| Float bindings in the given `Term` inwards, and calculate the set of
+        -- variable `Unique`s in the result `Term`.
+        go ::
+            Term tyname name uni fun a ->
+            Term tyname name uni fun (a, Uniques)
+        go t = case t of
+            Apply a fun0 arg0 ->
+                let fun = go fun0
+                    arg = go arg0
+                    us = termUniqs fun `Set.union` termUniqs arg
+                 in Apply (a, us) fun arg
+            LamAbs a n ty body0 ->
+                let body = go body0
+                 in LamAbs (a, termUniqs body) n (noUniq ty) body
+            TyAbs a n k body0 ->
+                let body = go body0
+                 in TyAbs (a, termUniqs body) n (noUniq k) body
+            TyInst a body0 ty ->
+                let body = go body0
+                 in TyInst (a, termUniqs body) body (noUniq ty)
+            IWrap a ty1 ty2 body0 ->
+                let body = go body0
+                 in IWrap (a, termUniqs body) (noUniq ty1) (noUniq ty2) body
+            Unwrap a body0 ->
+                let body = go body0
+                 in Unwrap (a, termUniqs body) body
+            Let a NonRec bs0 body0 ->
+                let bs = goBinding <$> bs0
+                    body = go body0
+                 in -- The bindings in `bs` should be processed from right to left, since
+                    -- a binding may depend on another binding to its left.
+                    runReader
+                        (foldrM (floatInBinding ver a) body bs)
+                        (FloatInContext False usgs)
+            Let a Rec bs0 body0 ->
+                -- Currently we don't move recursive bindings, so we simply descend into the body.
+                let bs = goBinding <$> bs0
+                    body = go body0
+                    us = Set.union (termUniqs body) (foldMap bindingUniqs bs)
+                 in Let (a, us) Rec bs body
+            Var a n -> Var (a, Set.singleton (n ^. PLC.unique)) n
+            Constant{} -> (,mempty) <$> t
+            Builtin{} -> (,mempty) <$> t
+            Error{} -> (,mempty) <$> t
+
+        -- \| Float bindings in the given `Binding` inwards, and calculate the set of
+        -- variable `Unique`s in the result `Binding`.
+        goBinding ::
+            Binding tyname name uni fun a ->
+            Binding tyname name uni fun (a, Uniques)
+        goBinding = \case
+            TermBind a s var rhs ->
+                let rhs' = go rhs
+                 in TermBind (a, termUniqs rhs') s (noUniq var) rhs'
+            other -> noUniq other
+
+termUniqs :: Term tyname name uni fun (a, Uniques) -> Uniques
+termUniqs = snd . termAnn
+
+-- | The result includes both the `Unique` of the bound var and the `Unique`s in the RHS.
```

Comment:
`Uniques` currently contains all variables, not just free variables, so for consistency reason I'm including the bound variables. Arguably we should calculate only the set of free variables since it could be much smaller.

Reply:
Updated the comment and also added a comment on `type Uniques = ...`